pyhf/pdf.py

```diff
@@ -70,68 +70,49 @@ def par_slice(self, name):
     def param_set(self, name):
         return self.par_map[name]['parset']
 
-    def modifier(self, name):
-        return self.par_map[name]['modifier']
-
     def set_poi(self,name):
-        if name not in self.modifiers:
+        if name not in [x for x,_ in self.modifiers]:
             raise exceptions.InvalidModel("The paramter of interest '{0:s}' cannot be fit as it is not declared in the model specification.".format(name))
         s = self.par_slice(name)
         assert s.stop-s.start == 1
         self.poi_index = s.start
 
-    def register_paramset(self, name, n_parameters, modifier):
-        '''allocates n nuisance parameters and stores paramset > modifier map'''
-        log.info('adding modifier %s (%s new nuisance parameters)', name, n_parameters)
+    def _bookkeep_paramset(self, channel, sample, name, modifier_type):
+        # get modifier class associated with modifier type
+        try:
+            modifier_cls = modifiers.registry[modifier_type]
+        except KeyError:
+            log.exception('Modifier type not implemented yet (processing {0:s}). Current modifier types: {1}'.format(modifier_type, modifiers.registry.keys()))
+            raise exceptions.InvalidModifier()
+
+        parset = modifier_cls.required_parset(len(sample['data']))
```

Comment:
seems like if we just pass the parset here

```
def _bookkeep_paramset(self, parset):
```

the APi would be cleaner (also see comment below, i'm not sure why we need sample / channel)

Reply:
right, but if we just craete the parset requirements elsewhere and pass it in, we don't anymore, right?